### DISCUSSION TITLE
Generate new slide from list of images

### DISCUSSION THREAD

**Discussion by baslat** [quarto-dev/quarto-cli#2403] [Q&A] [answered]
I am making a powerpoint presentation in quarto with R. I have a list of n images, and I have a list of n strings that describe the images. I would like each image to be on a new slide, accompanied by its description. How can I write code to dynamcially generate the slides?

It's similar to this (which I cannot get to work with quarto): https://stackoverflow.com/questions/11956520/r-knitr-markdown-output-plots-within-for-loop, and this (but not python): https://github.com/quarto-dev/quarto-cli/discussions/954

Thanks!

Here's an example of what I'm working with (note I had to add a comment to the trailing backticks on theh code blocks to escape the formatting).

```
---
title: "Untitled"
format: pptx
---

```{r}
library(tidyverse)
cyls <- unique(mtcars$cyl)

make_plot <- function(cyls) {
mtcars  |>
        filter(cyl == cyls) |>
        ggplot(aes(x = mpg, y = hp)) +
        geom_point() +
        labs(title = paste(cyls,"cylinders"))
}

plots <- purrr::map(cyls, make_plot)

descs <- list(
    "These cars are good",
    "These cars are fun",
    "I have no opinion on these cars")
```#neededasescape



<!-- Done manually -->

## 6 Cyls

`r descs[[1]]`

```{r}
print(plots[[1]])
```#neededasescape

## 4 Cyls

`r descs[[2]]`

```{r}
print(plots[[2]])
```#neededasescape


## 8 Cyls

`r descs[[3]]`

```{r}
print(plots[[3]])
```#neededasescape


```

**Comment by cderv** [comment, marked as answer]
I believe this answer regarding **gt** tables created dynamically can help understand how you could do that with **knitr**. This is linking to resources already working in R Markdown ecosystem. 

https://github.com/quarto-dev/quarto-cli/issues/2370#issuecomment-1241946395

Without your example, I would do that 

````markdown
---
title: "Untitled"
format: pptx
---

```{r, include=FALSE}
library(tidyverse)
cyls <- unique(mtcars$cyl)

descs <- c(
    "These cars are good",
    "These cars are fun",
    "I have no opinion on these cars")

```

```{r}
#| output: asis
src <- purrr::map2_chr(cyls, descs, \(cyls, desc) {
  knitr::knit_expand(text = c(
    "## {{ cyls }} Cyls",
    "",
    "{{ desc }}",
    "",
    "",
    "```{r}",
    "#| echo: true",
    "mtcars  |>",
    "  filter(cyl == {{ cyls }}) |>",
    "  ggplot(aes(x = mpg, y = hp)) +",
    "  geom_point() +",
    "  labs(title = paste(cyls,'cylinders'))",
    "```",
    ""))
})
res <- knitr::knit_child(text = src, quiet = TRUE)
cat(res, sep = '\n')
```
````

This will create 4 slides with the content you want to iterate on
![image](https://user-images.githubusercontent.com/6791940/189728830-94da2aa0-a20b-4010-8586-546c3fe0bacc.png)


This can be tweak by using `knit_child` only and using an external child document, or template to user. 

Hope it helps understand what **knitr** offers for this kind of usage.

**Reply by baslat** [reply]
Beautiful, thank yoo so much!

**Reply by baslat** [reply]
Do you know how I could modify this to use an existing list of ggplots, instead of the code to create the plots?

**Reply by baslat** [reply]
My hacky solution for the existing list is as follows (assuming `plots` is a list of 3 charts):



````
```{r}
plots <- list(chart1, chart2, chart3)
index <- 1:3
```


```{r make_charts}
#| output: asis
src <- purrr::map_chr(index, \(index) {
  knitr::knit_expand(text = c(
    "## Some title",
    "",
    "Words here, could be from a desc",
    "",
    "",
    "```{r}",
    "#| echo: false",
    "plots[[{{ index }}]]",
    "```",
    ""))
    }
    )
res <- knitr::knit_child(text = src, quiet = TRUE)
cat(res, sep = '\n')

```
````

**Reply by cderv** [reply]
Yes this works. For a list of object, I would not use `knit_expand()`, just `knit_child()`. 

````markdown
---
title: "Untitled"
format: pptx
keep-md: true
---

```{r}
library(tidyverse)
cyls <- unique(mtcars$cyl)

make_plot <- function(cyls) {
  mtcars  |>
    filter(cyl == cyls) |>
    ggplot(aes(x = mpg, y = hp)) +
    geom_point() +
    labs(title = paste(cyls,"cylinders"))
}

plots <- purrr::map(cyls, make_plot)

descs <- list(
    "These cars are good",
    "These cars are fun",
    "I have no opinion on these cars")

tabs <- tibble::tibble(
  cyls = cyls,
  plots = plots,
  descs = unlist(descs)
)
```


```{r}
#| output: asis
res <- purrr::pmap_chr(tabs, \(cyls, plots, descs) {
  knitr::knit_child(text = c(
    "## `r cyls` Cyls",
    "",
    "`r descs`",
    "",
    "",
    "```{r}",
    "#| echo: false",
    "plots",
    "```",
    ""), envir = environment(), quiet = TRUE)
})
cat(res, sep = '\n')
```
`````

I let you try to understand this example, and tweak it to your need.
